libtenzir/builtins/functions/numeric.cpp

```diff
@@ -37,17 +38,49 @@ class round_use final : public function_use {
     // round(x, 1h) -> round to multiples of 1h
     // round(<time>, <duration>)
     // round(x, 1h) -> round so that time is multiples of 1h (for UTC timezone?)
-    auto spec = located{eval(*spec_), spec_->get_location()};
-    return round_with_spec(std::move(value), std::move(spec), ctx);
+    return round_with_spec(std::move(value), std::move(spec_.value()), ctx);
   }
 
 private:
-  static auto round_with_spec(located<series> value, located<series> spec,
+  static auto round_with_spec(located<series> value, located<duration> spec,
                               session ctx) -> series {
-    diagnostic::warning("round(_, _) is not implemented yet")
-      .primary(spec)
-      .emit(ctx);
-    return series::null(value.inner.type, value.inner.length());
+    if (spec == decltype(spec){}) {
+      diagnostic::warning("resolution must not be 0").primary(spec).emit(ctx);
+      return series::null(value.inner.type, value.inner.length());
+    }
+    auto f = detail::overload{
+      [&](const arrow::DurationArray& array) -> series {
+        auto b
+          = duration_type::make_arrow_builder(arrow::default_memory_pool());
+        check(b->Reserve(array.length()));
+        for (auto i = int64_t{0}; i < array.length(); i++) {
+          if (array.IsNull(i)) {
+            check(b->AppendNull());
+            continue;
+          }
+          const auto val = array.Value(0);
+          const auto count = spec.inner.count();
+          const auto rem = val % std::abs(count);
+          const auto delta = rem * 2 < count ? -rem : count - rem;
+          check(b->Append(val + delta));
+        }
+        return {duration_type{}, finish(*b)};
+      },
+      [&](const arrow::TimestampArray& array) -> series {
+        auto rounded_array = arrow::compute::RoundTemporal(
+                               array, make_round_temporal_options(spec.inner))
+                               .ValueOrDie()
```

Comment:
Can we check the result of this with `check`? Killing the operator is better than killing the process.